### PR TITLE
built-in: add time-to-sec built-in function

### PR DESCRIPTION
**Pull request by Xuanwo**
Signed-off-by: Xuanwo <xuanwo.cn@gmail.com>

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -1965,7 +1966,28 @@ type builtinTimeToSecSig struct {
...
+	curTime := []int64{}
+	for _, v := range strings.Split(t, ":") {
+		int64V, err := strconv.ParseInt(v, 10, 64)
+		if err != nil {
+			return d, errors.Trace(err)
+		}
+		curTime = append(curTime, int64V)
+	}
+
+	d.SetInt64(curTime[0]*3600 + curTime[1]*60 + curTime[2])
```
Comment (shenli) [inline expression/builtin_time.go]: Please check the length of curTime and make sure curTime is a valid time.

File: expression/builtin_time_test.go [hunk 2]
```diff
@@ -1159,3 +1159,26 @@ func (s *testEvaluatorSuite) TestGetFormat(c *C) {
...
+	tests := []struct {
+		t      string
+		expect int64
+	}{
+		// Test case from https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_time-to-sec
+		{"22:23:00", 80580},
+		{"00:39:38", 2378},
+		// Test case for boundary values
+		{"00:00:00", 0},
+		{"23:59:59", 86399},
```
Comment (shenli) [inline expression/builtin_time_test.go]: How about "00:00" and "-1:-1:-1"?

**Comment by shenli** [comment]
@Xuanwo Thanks for your PR!
@coocood @XuHuaiyu PTAL

File: expression/builtin_time.go (hunk 1 above)
Comment (Xuanwo) [inline expression/builtin_time.go]: I changed to use `convertToDuration` to do this job, so I don't need to do that chenk any more.

File: expression/builtin_time_test.go (hunk 2 above)
Comment (Xuanwo) [inline expression/builtin_time_test.go]: I add a `00:00` test case, and `-2:-2:-2` will get a `invalid time format` error throw by ParseDuration.

Comment (XuHuaiyu) [inline expression/builtin_time_test.go]: @tiancaiamao 

In MySQL, for sql like `select TIME_TO_SEC('-2:-2:-2')`, we get a warning
```
  Level: Warning
   Code: 1292
Message: Truncated incorrect time value: '-2:-2:-2'
```

But in TiDB, the input will cause an error [here](https://github.com/pingcap/tidb/blob/master/util/types/time.go#L914)

File: expression/builtin_time.go [hunk 3]
```diff
@@ -1965,7 +1965,24 @@ type builtinTimeToSecSig struct {
 
 // See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_time-to-sec
 func (b *builtinTimeToSecSig) eval(row []types.Datum) (d types.Datum, err error) {
-	return d, errFunctionNotExists.GenByArgs("TIME_TO_SEC")
+	args, err := b.evalArgs(row)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	d, err = convertToDuration(b.ctx.GetSessionVars().StmtCtx, args[0], types.MaxFsp)
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: why we need types.MaxFsp here? It seems fsp has no effect on the function.

File: expression/builtin_time.go [hunk 4]
```diff
@@ -1965,7 +1965,24 @@ type builtinTimeToSecSig struct {
...
+	if err != nil || d.IsNull() {
+		return d, errors.Trace(err)
+	}
+
+	t := d.GetMysqlDuration()
+	if t.Compare(types.ZeroDuration) < 0 {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: Add a *TODO* comment here for the cases like `select TIME_TO_SEC('-2:-2:-2')`

File: expression/builtin_time_test.go [hunk 5]
```diff
@@ -1159,3 +1159,33 @@ func (s *testEvaluatorSuite) TestGetFormat(c *C) {
 		c.Assert(result, Equals, test.expect)
 	}
 }
+
+func (s *testEvaluatorSuite) TestTimeToSec(c *C) {
+	tests := []struct {
+		t      string
+		expect int64
+	}{
+		// Test case from https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_time-to-sec
```
Comment (XuHuaiyu) [inline expression/builtin_time_test.go]: Remove this comment.

File: expression/builtin_time_test.go [hunk 6]
```diff
@@ -1159,3 +1159,33 @@ func (s *testEvaluatorSuite) TestGetFormat(c *C) {
...
+		{"22:23:00", 80580},
+		{"00:39:38", 2378},
+		// Test case for boundary values
+		{"23:00", 82800},
+		{"00:00", 0},
+		{"00:00:00", 0},
+		{"23:59:59", 86399},
+		// Test case for not normal values
```
Comment (XuHuaiyu) [inline expression/builtin_time_test.go]: ditto.

File: expression/builtin_time_test.go [hunk 7]
```diff
@@ -1159,3 +1159,33 @@ func (s *testEvaluatorSuite) TestGetFormat(c *C) {
...
+
+func (s *testEvaluatorSuite) TestTimeToSec(c *C) {
+	tests := []struct {
+		t      string
+		expect int64
+	}{
+		// Test case from https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_time-to-sec
+		{"22:23:00", 80580},
+		{"00:39:38", 2378},
+		// Test case for boundary values
```
Comment (XuHuaiyu) [inline expression/builtin_time_test.go]: ditto.

File: expression/builtin_time_test.go (hunk 6 above)
Comment (Xuanwo) [inline expression/builtin_time_test.go]: all removed.

File: expression/builtin_time.go (hunk 3 above)
Comment (Xuanwo) [inline expression/builtin_time.go]: I'm not sure, code before use types.MaxFsp mostly, and I'll set it to zero.

File: expression/builtin_time.go (hunk 4 above)
Comment (Xuanwo) [inline expression/builtin_time.go]: Added.

**Comment by Xuanwo** [comment]
@XuHuaiyu @shenli Please see the latest commit.

**Comment by Xuanwo** [comment]
@XuHuaiyu @shenli Hi, any more review?

**Comment by shenli** [comment]
LGTM @Xuanwo Please resolve the conflicts.
@XuHuaiyu PTAL

**Comment by Xuanwo** [comment]
@shenli  conflicts solved.

File: expression/builtin_time.go [hunk 8]
```diff
@@ -1965,7 +1965,25 @@ type builtinTimeToSecSig struct {
...
+		return d, errors.Trace(err)
+	}
+
+	d, err = convertToDuration(b.ctx.GetSessionVars().StmtCtx, args[0], 0)
+	if err != nil || d.IsNull() {
+		return d, errors.Trace(err)
+	}
+
+	t := d.GetMysqlDuration()
+	// TODO: select TIME_TO_SEC('-2:-2:-2') not handle well
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: Add `.` at the end of this comment.

Comment (Xuanwo) [inline expression/builtin_time.go]: Added.

**Comment by Xuanwo** [comment]
@XuHuaiyu Done!

**Pull request merged**